File: src/main/java/com/perforce/performance/config/ConfigReader.java
package com.perforce.performance.config;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

/**
 * ConfigReader - Reads application configuration from config.properties
 * Update config.properties with your app URL, credentials, and screen paths.
 */
public class ConfigReader {

    private static final Logger log = LoggerFactory.getLogger(ConfigReader.class);
    private static final Properties properties = new Properties();
    private static ConfigReader instance;

    private ConfigReader() {
        loadProperties();
    }

    public static ConfigReader getInstance() {
        if (instance == null) {
            instance = new ConfigReader();
        }
        return instance;
    }

    private void loadProperties() {
        try (InputStream input = getClass().getClassLoader().getResourceAsStream("config.properties")) {
            if (input == null) {
                throw new RuntimeException("config.properties not found in classpath");
            }
            properties.load(input);
            log.info("Configuration loaded successfully");
        } catch (IOException e) {
            throw new RuntimeException("Failed to load config.properties: " + e.getMessage(), e);
        }
    }

    public String get(String key) {
        String value = properties.getProperty(key);
        if (value == null) {
            throw new RuntimeException("Property not found: " + key);
        }
        return value.trim();
    }

    public String get(String key, String defaultValue) {
        return properties.getProperty(key, defaultValue).trim();
    }

    public int getInt(String key, int defaultValue) {
        try {
            return Integer.parseInt(get(key, String.valueOf(defaultValue)));
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    public boolean getBoolean(String key, boolean defaultValue) {
        return Boolean.parseBoolean(get(key, String.valueOf(defaultValue)));
    }

    // ─── Convenience getters ───────────────────────────────────────────────────

    public String getBaseUrl()         { return get("app.base.url"); }
    public String getLoginUrl()        { return get("app.login.url", getBaseUrl() + "/login"); }
    public String getUsername()        { return get("app.username"); }
    public String getPassword()        { return get("app.password"); }
    public String getBrowser()         { return get("browser", "chrome"); }
    public boolean isHeadless()        { return getBoolean("browser.headless", false); }
    public int getPageLoadTimeout()    { return getInt("timeout.pageload", 30); }
    public int getImplicitWait()       { return getInt("timeout.implicit", 10); }
    public int getPerformanceRuns()    { return getInt("performance.runs", 3); }
    public String getReportPath()      { return get("report.output.path", "reports"); }
    public String getScreenshotPath()  { return get("report.screenshot.path", getReportPath() + "/screenshots"); }
}

File: src/config/ConfigReader.ts
import PropertiesReader from 'properties-reader';
import * as path from 'path';
import * as fs from 'fs';

/**
 * ConfigReader — singleton that loads config.properties from the project root.
 * Drop-in equivalent of the Java ConfigReader.
 */
export class ConfigReader {
  private static instance: ConfigReader;
  private reader: PropertiesReader.Reader;

  private constructor() {
    // Look for config.properties in the working directory (project root)
    const candidates = [
      path.resolve(process.cwd(), 'config.properties'),
      path.resolve(__dirname, '../../config.properties'),
      path.resolve(__dirname, '../../../config.properties'),
    ];
    const found = candidates.find(p => fs.existsSync(p));
    if (!found) {
      throw new Error(
        `config.properties not found. Tried:\n  ${candidates.join('\n  ')}`
      );
    }
    this.reader = PropertiesReader(found);
  }

  static getInstance(): ConfigReader {
    if (!ConfigReader.instance) {
      ConfigReader.instance = new ConfigReader();
    }
    return ConfigReader.instance;
  }

  get(key: string): string {
    const value = this.reader.get(key);
    if (value === null || value === undefined) {
      throw new Error(`Property not found: ${key}`);
    }
    return String(value).trim();
  }

  getOrDefault(key: string, defaultValue: string): string {
    const value = this.reader.get(key);
    if (value === null || value === undefined) return defaultValue;
    return String(value).trim();
  }

  getInt(key: string, defaultValue: number): number {
    const raw = this.getOrDefault(key, String(defaultValue));
    const parsed = parseInt(raw, 10);
    return isNaN(parsed) ? defaultValue : parsed;
  }

  getBoolean(key: string, defaultValue: boolean): boolean {
    return this.getOrDefault(key, String(defaultValue)).toLowerCase() === 'true';
  }

  // ── Convenience getters ────────────────────────────────────────────────────

  getBaseUrl(): string      { return this.get('app.base.url'); }
  getLoginUrl(): string     { return this.getOrDefault('app.login.url', this.getBaseUrl() + '/login'); }
  getUsername(): string     { return this.get('app.username'); }
  getPassword(): string     { return this.get('app.password'); }
  getBrowser(): string      { return this.getOrDefault('browser', 'chrome'); }
  isHeadless(): boolean     { return this.getBoolean('browser.headless', false); }
  getPageLoadTimeout(): number { return this.getInt('timeout.pageload', 30); }
  getImplicitWait(): number    { return this.getInt('timeout.implicit', 10); }
  getPerformanceRuns(): number { return this.getInt('performance.runs', 3); }
  getReportPath(): string      { return this.getOrDefault('report.output.path', 'reports'); }
  getScreenshotPath(): string  {
    return this.getOrDefault('report.screenshot.path', this.getReportPath() + '/screenshots');
  }
}
